csharp/src/Ice/Reference.cs

```diff
@@ -1524,192 +1318,138 @@ private void ClearConnection(Connection connection)
+                if (Protocol == Protocol.Ice1 && PreferNonSecure == NonSecure.Never && !endpoint.IsAlwaysSecure)
                 {
-                    throw new NoEndpointException(ToString());
+                    return false;
                 }
 
-                // Apply overrides and filter endpoints
-                IEnumerable<Endpoint> filteredEndpoints = endpoints.Where(endpoint =>
+                // Check if the endpoint is compatible with the proxy invocation mode, Twoway and Oneway invocation
+                // modes require a non datagram endpoint, Datagram invocation mode requires a datagram endpoint, the
+                // other invocation modes (BatchOneway and BatchDagram) are not supported.
+                return InvocationMode switch
                 {
-                    // Filter out opaque endpoints
-                    if (endpoint is OpaqueEndpoint || endpoint is UniversalEndpoint)
-                    {
-                        return false;
-                    }
-
-                    // Filter out based on InvocationMode and IsDatagram
-                    switch (InvocationMode)
-                    {
-                        case InvocationMode.Twoway:
-                        case InvocationMode.Oneway:
-                        case InvocationMode.BatchOneway:
-                            if (endpoint.IsDatagram)
-                            {
-                                return false;
-                            }
-                            break;
+                    InvocationMode.Twoway or InvocationMode.Oneway => !endpoint.IsDatagram,
+                    InvocationMode.Datagram => endpoint.IsDatagram,
+                    _ => false
+                };
+            });
 
-                        case InvocationMode.Datagram:
-                        case InvocationMode.BatchDatagram:
-                            if (!endpoint.IsDatagram)
-                            {
-                                return false;
-                            }
-                            break;
+            var orderedEndpoints = Communicator.OrderEndpointsByTransportFailures(filteredEndpoints).ToList();
+            if (orderedEndpoints.Count == 0)
+            {
+                throw new NoEndpointException(ToString());
+            }
+            return (orderedEndpoints, cached);
+        }
+        internal Connection? GetCachedConnection() => _connection;
 
-                        default:
-                            Debug.Assert(false);
-                            return false;
+        internal async ValueTask<Connection> GetConnectionAsync(CancellationToken cancel)
+        {
+            var linkedCancellationSource = CancellationTokenSource.CreateLinkedTokenSource(
+                cancel,
+                Communicator.CancellationToken);
+            cancel = linkedCancellationSource.Token;
+            Connection? connection = _connection;
```

Comment:
It would be better to check if the connection is non-null and active here to return it and move the creation of the cancellation token source after.

Reply:
Fixed